File: src/main/java/pe/nanamochi/banchus/packets/client/handlers/LobbyPartHandler.java
package pe.nanamochi.banchus.packets.client.handlers;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Set;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import pe.nanamochi.banchus.entities.db.Channel;
import pe.nanamochi.banchus.entities.db.Session;
import pe.nanamochi.banchus.entities.redis.PacketBundle;
import pe.nanamochi.banchus.packets.AbstractPacketHandler;
import pe.nanamochi.banchus.packets.PacketWriter;
import pe.nanamochi.banchus.packets.Packets;
import pe.nanamochi.banchus.packets.client.LobbyPartPacket;
import pe.nanamochi.banchus.packets.server.ChannelAvailablePacket;
import pe.nanamochi.banchus.services.auth.SessionService;
import pe.nanamochi.banchus.services.communication.ChannelMembersService;
import pe.nanamochi.banchus.services.communication.ChannelService;
import pe.nanamochi.banchus.services.protocol.PacketBundleService;

@Component
@RequiredArgsConstructor
public class LobbyPartHandler extends AbstractPacketHandler<LobbyPartPacket> {
  private static final Logger logger = LoggerFactory.getLogger(LobbyPartHandler.class);

  private final PacketWriter packetWriter;
  private final PacketBundleService packetBundleService;
  private final SessionService sessionService;
  private final ChannelService channelService;
  private final ChannelMembersService channelMembersService;

  @Override
  public Packets getPacketType() {
    return Packets.OSU_LOBBY_PART;
  }

  @Override
  public Class<LobbyPartPacket> getPacketClass() {
    return LobbyPartPacket.class;
  }

  @Override
  public void handle(LobbyPartPacket packet, Session session, ByteArrayOutputStream responseStream)
      throws IOException {
    logger.debug("Handling packet: {}", getPacketType());
    session.setReceiveMatchUpdates(false);
    sessionService.updateSession(session);

    Channel channel = channelService.findByName("#lobby");

    if (channel == null) return;

    channelMembersService.removeMemberFromChannel(channel, session);

    Set<UUID> currentChannelMembers = channelMembersService.getMembers(channel.getId());

    // TODO: Only get all sessions that has any privilege bit
    for (Session otherOsuSession : sessionService.getAllSessions()) {
      if (!channelService.canReadChannel(channel, session.getUser().getPrivileges())) continue;
      ByteArrayOutputStream stream = new ByteArrayOutputStream();
      packetWriter.writePacket(
          stream,
          new ChannelAvailablePacket(
              channel.getName(), channel.getTopic(), currentChannelMembers.size()));
      packetBundleService.enqueue(otherOsuSession.getId(), new PacketBundle(stream.toByteArray()));
    }

    logger.info("User {} has part from #lobby.", session.getUser().getUsername());
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/handlers/LobbyPartHandler.kt
package pe.nanamochi.banchus.handlers

import java.io.ByteArrayOutputStream
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import pe.nanamochi.banchus.core.PacketType
import pe.nanamochi.banchus.database.entity.Session
import pe.nanamochi.banchus.infrastructure.protocol.HandleClientPacket
import pe.nanamochi.banchus.packets.client.LobbyPartPacket
import pe.nanamochi.banchus.protocol.AbstractPacketHandler
import pe.nanamochi.banchus.service.ChannelService
import pe.nanamochi.banchus.service.SessionService

@Component
@HandleClientPacket(type = PacketType.OSU_LOBBY_PART)
class LobbyPartHandler(
    private val sessionService: SessionService,
    private val channelService: ChannelService,
) : AbstractPacketHandler<LobbyPartPacket>(PacketType.OSU_LOBBY_PART) {

    private val log = LoggerFactory.getLogger(javaClass)

    override fun handle(
        packet: LobbyPartPacket,
        session: Session,
        responseStream: ByteArrayOutputStream,
    ) {
        session.receiveMatchUpdates = false
    }
}
